### PR TITLE
Framelifecycle polishes (Updated Perfetto)

### PR DESCRIPTION
**Pull request by silence-do-good**
Now that the frame events are separated from the gpu_slice table, this PR addresses that change and the following:
- Show statistics of all the frames/buffers that got composited into a displayed frame, when it is selected
- Highlight(draw in color) all related events of the frame when an instant event is selected, greying out other events
- Highlight(draw in color) all related events of all the frames within a displayed frame slice when it is selected, greying out other events
- Minor: Make sure that displayed frame track always goes on top of all other tracks under the SF events group
- Minor: Remove trackId from multiselection view and show layer name instead

Bug: 152533807, 150172851

### CODE REVIEW

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 1]
```diff
@@ -333,6 +388,62 @@ public void getRange(Consumer<TimeSpan> span) {
...
+    public final List<String> layerNames;
+    public static FrameSelection EMPTY = new FrameSelection();
+
+    public FrameSelection() {
+      frameNumbers = Lists.newArrayList();
+      layerNames = Lists.newArrayList();
+    }
+
+    public FrameSelection(long[] frameNumbers, String[] layerNames) {
+      this.frameNumbers = Lists.newArrayList(Longs.asList(frameNumbers));
```
Comment (rschiu) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:410]: Just Lists.newArrayList?

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 2]
```diff
@@ -333,6 +388,62 @@ public void getRange(Consumer<TimeSpan> span) {
...
+      return false;
+    }
+
+    public void combine(FrameSelection other) {
+      frameNumbers.addAll(other.frameNumbers);
+      layerNames.addAll(other.layerNames);
+    }
+
+    public boolean isEmpty() {
+      return frameNumbers.size() == 0 ? true : false;
```
Comment (rschiu) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:430]: return frameNumbers.isEmpty();

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 3]
```diff
@@ -333,6 +388,62 @@ public void getRange(Consumer<TimeSpan> span) {
...
+    }
+
+    public FrameSelection(long[] frameNumbers, String[] layerNames) {
+      this.frameNumbers = Lists.newArrayList(Longs.asList(frameNumbers));
+      this.layerNames = Lists.newArrayList(layerNames);
+    }
+
+    public boolean contains(long[] f, String[] l) {
+      for (int i = 0; i < f.length; i++) {
+        int idx = frameNumbers.indexOf(f[i]);
```
Comment (rschiu) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:416]: This doesn't work if FrameSelection has multiple layers with the same frame number.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java [hunk 4]
```diff
@@ -403,6 +408,16 @@ public void computeSelection(CombiningBuilder builder, Area area, TimeSpan ts) {
     }
   }
 
+  private static FrameSelection getSelectedFrames(State state) {
+    Selection selection = state.getSelection(Selection.Kind.FrameEvents);
+    if (selection instanceof FrameEventsTrack.Slice) {
+      return ((FrameEventsTrack.Slice) selection).getSelection();
+    } else if (selection instanceof FrameEventsTrack.Slices) {
```
Comment (rschiu) [inline gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:415]: Can you combine FrameEventsTrack.Slice with FrameEventsTrack.Slices?  Stella did something similar.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java [hunk 5]
```diff
@@ -323,14 +329,13 @@ private Hover sliceHover(FrameEventsTrack.Data data, Fonts.TextMeasurer m, doubl
         endts = ts + 12;
         ts -= 12;
       }
-      if (data.depths[i] == depth && x >= ts && x<= endts) {
+      if (x >= ts && x<= endts) {
```
Comment (rschiu) [inline gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:332]: Same question with comment above.  Is this change necessary?

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 6]
```diff
@@ -160,7 +196,7 @@ private static String sliceSql(long id) {
 
   public ListenableFuture<List<Slice>> getSlices(TimeSpan ts, int minDepth, int maxDepth) {
     return transform(qe.query(sliceRangeSql(ts, minDepth, maxDepth)),
-        res -> res.list(($, row) -> buildSlice(row, ArgSet.EMPTY)));
+        res -> res.list(($, row) -> buildSlice(row, ArgSet.EMPTY, null)));
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:199]: Nit: personal opinion but I think an empty map might be better. The same for null below.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 7]
```diff
@@ -205,14 +241,14 @@ public Data(DataRequest request, long bucketSize, String[] concatedIds, long[] n
       this.ids = null;
       this.starts = null;
       this.ends = null;
-      this.depths = null;
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:208]: The depth is still queried as one of the base columns in the SQL query, and we also still keep the getSlices method inside this file using depth as parameter.. Do we want to remove that as well?

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 8]
```diff
@@ -333,6 +388,62 @@ public void getRange(Consumer<TimeSpan> span) {
...
+      FrameSelection selection = new FrameSelection();
+      slices.forEach(s -> selection.combine(s.getSelection()));
+      return selection;
+    }
+  }
+
+  public static class FrameSelection {
+    public final List<Long> frameNumbers;
+    public final List<String> layerNames;
+    public static FrameSelection EMPTY = new FrameSelection();
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:402]: nit: would be better to put it at the first line of the class.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 9]
```diff
@@ -291,6 +343,9 @@ public void getRange(Consumer<TimeSpan> span) {
       }
     }
 
+    public FrameSelection getSelection() {
+      return new FrameSelection(frameNumbers, layerNames);
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:347]: This function will be queried every time rendering the track, so it would be good to keep it efficient. It seems that frameNumbers and layerNames are fixed at slice creation time, maybe we can save a fixed FrameSelection at slice creation time as well? So as to avoid creating a new object every time. Or maybe I'm missing something here.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSelectionView.java [hunk 10]
```diff
@@ -56,6 +59,34 @@ public FrameEventsSelectionView(Composite parent, State state, FrameEventsTrack.
...
+            withSpans(new GridData(), 2, 1));
+
+        createLabel(stats, "Queue to Acquire: ");
+        createLabel(stats, timeToString(v.queueToAcquireTime));
+
+        createLabel(stats, "Acquire to Latch: ");
+        createLabel(stats, timeToString(v.acquireToLatchTime));
+
+        createLabel(stats, "Latch to Present: ");
+        createLabel(stats, timeToString(v.latchToPresentTime));
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSelectionView.java:80]: Can we show frameNumbers as well here? I think we should show users a little hint why something got grouped together by color. 
From the code I see they are connected by frameNumbers and layerNames, but these things are not shown in UI.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 11]
```diff
@@ -333,6 +388,62 @@ public void getRange(Consumer<TimeSpan> span) {
...
+      layerNames = Lists.newArrayList();
+    }
+
+    public FrameSelection(long[] frameNumbers, String[] layerNames) {
+      this.frameNumbers = Lists.newArrayList(Longs.asList(frameNumbers));
+      this.layerNames = Lists.newArrayList(layerNames);
+    }
+
+    public boolean contains(long[] f, String[] l) {
+      for (int i = 0; i < f.length; i++) {
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:415]: Just a note that this function will also be queried every time rendering the track, and will be called for every visible slices. Is there any better key to identify the selection, and make this function run faster? If not then it's fine to keep it this way.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java [hunk 12]
```diff
@@ -176,10 +176,16 @@ public void renderSlices(RenderContext ctx, FrameEventsTrack.Data data) {
...
+        ctx.setBackgroundColor(color.disabled);
+        if (selectedFrames.contains(data.frameNumbers[i], data.layerNames[i])) {
+          color.applyBase(ctx);
+        }
+      }
 
       if (tEnd - tStart > 3 ) {
         double rectWidth = Math.max(1, state.timeToPx(tEnd) - rectStart);
-        double y = depth * SLICE_HEIGHT;
+        double y = 0;
```
[the remark this answers is not in the document]
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:188]: Removed the depth from the query altogether because its just not being used anywhere. And currently, there is no plan of using it in the foreseeable future as well.  Even if we change the instant events to slices, the depth will always be 0. We can always add it back in case it is needed, but its just a waste for now, with so many select queries being sent.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 3 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:416]: The lines below this does that check

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 7 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:208]: My bad. I removed the depth from the query as well, but I think that got lost while cleaning up. I'll remove the depth from the queries altogether.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java (hunk 4 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:415]: I'll look into what Stella did and make changes if its suitable here.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 9 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:347]: This function will be called only when the event is selected. Storing the FrameSelection within the Slice would be a waste because the slices are created anew for every redraw.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 3 above)
Comment (rschiu) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:416]: I am thinking of the scenario where the layer name comes at a later index.  indexOf only returns the first one, the function will return false instead of true.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 9 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:347]: Hmmm from my understanding, if following the chain of FrameEventSummaryPanel.java -> renderTrack -> renderSlices -> getSelectedFrames -> getSelection -> new FrameSelection, it's called at every track rendering time?

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 7 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:208]: Just checked again, depth is being used in the span_join operator and removing it from the query is going to change a lot of things. I suggest just removing depths from the data for now, since we are not using it.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java (hunk 4 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:415]: To avoid type checking here, it will require `FrameEventsTrack.Slice` and `FrameEventsTrack.Slices` both implement `Selection` framework, and have them both saved under a new category `Kind.FurtherFrame`.
I can see the difficulty of letting them implementing Selection framework here, because the class's contains signature is `contains(long[] f, String[] l)` rather than `contains(long id)`, if we really can't find a long-typed id as a key, I'm fine with it checked in as this for now.
There are now two outliners doing type checking, one her and one in VulkanEvent track. We'd better think about how to add some flexibility into the selection framework to make these two parts more elegant.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 11 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:417]: Leaving it like this for now, based on our discussion.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSelectionView.java (hunk 10 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSelectionView.java:83]: Done.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 9 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java]: Got it. Done.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 6 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:196]: We check for null in the selection view and its kinda easier this way. There is a solid difference between an instant event and a displayed frame slice.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java (hunk 5 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:331]: Discussed in our meeting about this and keeping it as is.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java (hunk 3 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java]: Fixed it. Done.

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java (hunk 4 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSummaryPanel.java:414]: As Stella pointed out, grouping them together is going to involve a lot of changes to the Selection framework and I'd like to avoid mixing those things up in the change. Stella and I will look into refactoring the Selection framework further, if we stumble upon this usecase again. Right now, only the Vulkan events and Frame events require this behavior.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Data about a GPU in the trace.
+ */
+public class FrameInfo {
+  public static final FrameInfo NONE = new FrameInfo(Collections.emptyList());
+
+  private static final String FRAME_SLICE_QUERY =
+      "select t.id, t.name, t.scope, max(depth) + 1 " +
+      "from gpu_track t join frame_slice s on (t.id = s.track_id) " +
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java:36]: This doesn't look right. Shouldn't this be limited to a different table than gpu_track?

File: gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        transform(qe.queries(FRAME_SLICE_QUERY), frame -> {
+            List<Buffer> buffers = Lists.newArrayList();
+            frame.forEachRow(($, r) -> {
+              buffers.add(new Buffer(r));
+            });
+
+            // Sort buffers by name, the query is sorted by track id for the queues.
+            buffers.sort((b1, b2) -> {
+              // Displayed Frame track should always be at top
+              if (b1.name.equals("Displayed Frame")) {
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java:73]: Yikes, sounds like the buffer should have some type enum. At least make this a constant.

File: gapic/src/main/com/google/gapid/perfetto/models/Tracks.java [hunk 15]
```diff
@@ -180,6 +167,20 @@ private Tracks() {
...
+      String parent = "sf_events";
+      for (FrameInfo.Buffer buffer : data.getFrame().buffers()) {
+        FrameEventsTrack track = FrameEventsTrack.forBuffer(data.qe, buffer);
+        data.tracks.addTrack(parent, track.getId(), buffer.getDisplay(),
+            single(state -> new FrameEventsSummaryPanel(state, buffer, track), true, false));
+      }
+    }
+    return data;
+  }
+
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/Tracks.java:183]: These tracks should still be in the gpu group, no?

Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/Tracks.java:183]: Moved them out of that group since the tables are now different as well. They will be just below the GPU for now.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java (hunk 13 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java:36]: Good catch actually. I think moving this out of gpu_track is going to involve a bit of changes in perfetto judging by the places its being touched. I'll create a follow up bug and once that is done, I'll patch this separately? This change has some important UI changes that I want to get in first.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java (hunk 14 above)
Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java]: Yeah. There are some potential changes coming along the way to the Displayed Frame track.  I'll refactor this logic depending on the decisions we make. For now, I've changed the literal into a constant.

File: gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java [hunk 16]
```diff
@@ -333,6 +391,59 @@ public void getRange(Consumer<TimeSpan> span) {
...
+    public FrameSelection(long[] f, String[] l) {
+      keys = Lists.newArrayList();
+      for (int i = 0; i < l.length; i++) {
+        keys.add(l[i] + "_" + f[i]);
+      }
+    }
+
+    public boolean contains(long[] f, String[] l) {
+      for (int i = 0; i < f.length; i++) {
+        if (keys.contains(l[i] + "_" + f[i])) {
```
Comment (stellama0208) [inline gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:418]: It looks like we can change `List<String> keys` to `Set<String> keys` to improve efficiency this line. Or maybe I'm missing sth and there's a reason using list here?

Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/models/FrameEventsTrack.java:418]: Done.
